File: C Programming/SortingAlgorithm/heapSort.c
/*
 * heapSort.c
 *
 *  Created on: Aug 12, 2018
 *      Author: prabhaka
 */
/*
 * 1. build a max heap
 * 2. move large element to end of heap , reduce size of heap.
 * 3. heapify on root again and repeat step 2 till size becomes 1
 *
 * */
int getParent(int i)
{
	return (i-1)/2;
}
int getLeft(int i)
{
	return (i *2) + 1;
}
int getRight(int i)
{
	return (i *2) + 2;
}
void swap(int *a, int *b)
{
	 int t = *a;
	 *a = *b;
	 *b = t;
}
void heapify(int inp[], int n, int root)
{

    if(root >= n)
    	return;
    int largest = root;
    int left = getLeft(root);
    int right = getRight(root);
    if(left < n && inp[largest] < inp[left])
    {
    	largest = left;
    }
    if(right < n && inp[largest] < inp[right])
    {
       largest = right;
    }
    if(largest != root)
    {
    	swap(&inp[root], &inp[largest]);
    	heapify(inp,n,largest);
    }

}

void heapSort(int inp[], int n)
{
	//heapify the array
	// to heapify any array start from the last parent and heapify up recursively
    int i = n/2-1;
    for(;i>=0; i--)
    {
    	heapify(inp,n,i);
    }
    printArray(inp,n);
    for(i = n-1; i >= 0 ; i--)
    {
    	swap(&inp[i], &inp[0]);
    	heapify(inp,i,0);

    }

}

File: PythonProgramming/Algo/sorting/heap_sort.py
# -*- coding: utf-8 -*-
"""
Created on Thu Feb 15 05:11:03 2024

@author: sanka
"""

# Heap sort
# Array will have two parts sorted and unsorted parts.
# array will have unsorted part at the end of it
# 
# Step 1 : Heapify from root
# Step 2 : Remove root to storted part of array
# Step 3 : Replace root with last element from unsorted and repeat from  Step1 
#          till arry is completely sorted

class heapsort:
    def __init__(self):
        pass
    def sort(self, array):
        self.print_array()(array)
        for i in range(0,len(array)):
            self.heapify_top_down(array, len(array), 0)
        self.print_array(array)
        
        
        pass
    
    def heapify_top_down(self, array,length ,start_index):
        largest = start_index
        left_child = self.getleftchild(largest)
        right_child = self.getrightchild(largest)
        
        if largest < length and array[left_child] > array[largest]:
            largest = left_child
        if largest < length and array[right_child] > array[largest]:
            largest = right_child
        ## Swap largest with 
        if start_index != largest:
            array[start_index], array[largest] = array[largest], array[start_index]
            self.heapify_top_down(array, length, largest)
    
    def print_array(self,array):
        print(array)
        
    def getrightchild(self, n):
        return 2*n + 1
    def getleftchild(self,n):
        return 2*n+2
    def getparent(self,n):
        return ( n - 1 ) //  2